thrust/thrust/detail/reference.h

```diff
@@ -179,10 +176,7 @@ class reference
   // about what system the object is on.
   _CCCL_HOST_DEVICE operator value_type() const
   {
-    // Avoid default-constructing a system; instead, just use a null pointer
-    // for dispatch. This assumes that `get_value` will not access any system
-    // state.
-    typename thrust::iterator_system<pointer>::type* system = nullptr;
+    typename thrust::iterator_system<pointer>::type system;
```

Comment:
I honestly would also prefer if we did not need to construct systems. From the recent CI failures and the generally open design (users can bring their own systems) we have to assume that users could rely on Thrust not trying to default construct a system. Given the current code "worked" and no data of the system was actually accessed, we can assume that only the type information should be needed. I assume/hope there is a better solution.

Reply:
I wondered how `thrust::device_vector` produces a system when it inits the elements during construction: It tries to get the system from the allocator. But in case the allocator has no system, it default constructs a system (thrust knows the type of the system for an allocator): https://github.com/NVIDIA/cccl/blob/87f72468ba4a094ab03d0f901d670c261a2e9e6e/thrust/thrust/detail/allocator/allocator_traits.inl#L270-L284

So it seems we have precedence for default constructing systems when needed.